sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -2685,67 +2768,156 @@ atomic_update(Tx *p, Toffset offset, simd<Tx, N> src0, simd_mask<N> mask) {
 /// has no arguments in addition to the value at the memory location.
 ///
 /// @tparam Op The atomic operation - can be \c atomic_op::inc or
-/// \c atomic_op::dec, \c atomic_op::load.
-/// @tparam Tx The vector element type.
+///   \c atomic_op::dec, \c atomic_op::load.
+/// @tparam T The vector element type.
 /// @tparam N The number of memory locations to update.
 /// @param p The USM pointer.
-/// @param offset The vector of 32-bit or 64-bit offsets in bytes.
+/// @param offset The vector of 32-bit or 64-bit offsets in bytes (zero-based).
 /// @param mask Operation mask, only locations with non-zero in the
 ///   corresponding mask element are updated.
+/// @param props The parameter 'props' specifies the optional compile-time
+///   properties list. Only L1/L2 properties is used.
+//    Other properties are ignored.
 /// @return A vector of the old values at the memory locations before the
 ///   update.
 ///
-template <atomic_op Op, typename Tx, int N, typename Toffset>
-__ESIMD_API simd<Tx, N> atomic_update(Tx *p, simd<Toffset, N> offset,
-                                      simd_mask<N> mask) {
-  static_assert(std::is_integral_v<Toffset>, "Unsupported offset type");
-  if constexpr (Op == atomic_op::load) {
-    if constexpr (std::is_integral_v<Tx>) {
-      return atomic_update<atomic_op::bit_or, Tx, N>(p, offset, simd<Tx, N>(0),
-                                                     mask);
-    } else {
-      using Tint = detail::uint_type_t<sizeof(Tx)>;
-      simd<Tint, N> Res = atomic_update<atomic_op::bit_or, Tint, N>(
-          reinterpret_cast<Tint *>(p), offset, simd<Tint, N>(0), mask);
-      return Res.template bit_cast_view<Tx>();
-    }
+template <atomic_op Op, typename T, int N, typename Toffset,
+          typename PropertyListT =
+              ext::oneapi::experimental::detail::empty_properties_t>
+__ESIMD_API std::enable_if_t<
+    __ESIMD_DNS::get_num_args<Op>() == 0 &&
+        ext::oneapi::experimental::is_property_list_v<PropertyListT>,
+    __ESIMD_NS::simd<T, N>>
+atomic_update(T *p, simd<Toffset, N> offset, simd_mask<N> mask,
+              PropertyListT props = {}) {
+  constexpr auto L1Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L1_key>(
+          cache_hint::none);
+
+  constexpr auto L2Hint =
+      detail::getPropertyValue<PropertyListT, cache_hint_L2_key>(
+          cache_hint::none);
+
+  static_assert(!PropertyListT::template has_property<cache_hint_L3_key>(),
+                "L3 cache hint is reserved. The old/experimental L3 LSC cache "
+                "hint is cache_level::L2 now.");
+
+  if constexpr (L1Hint != cache_hint::none || L2Hint != cache_hint::none ||
+                !__ESIMD_DNS::isPowerOf2(N, 32)) {
+    return detail::atomic_update_impl<Op, T, N, L1Hint, L2Hint, Toffset>(
+        p, offset, mask);
   } else {
-    detail::check_atomic<Op, Tx, N, 0>();
+    static_assert(std::is_integral_v<Toffset>, "Unsupported offset type");
```

Comment:
Should we move this assert outside the ifs? We always want the offset type to be integral right?

Reply:
Done in https://github.com/intel/llvm/pull/11833/commits/ffe19702c45355afd8b7ab17d2170cd783e9f670.